### ISSUE TITLE
Blog plugin to not mix on_files and on_nav objects

### ISSUE THREAD

**Issue by ultrabug** [squidfunk/mkdocs-material#5909]
### Context

I'm working on making sure the `mkdocs-static-i18n` plugin is compatible with the blog plugin

### Description

The blog plugin adds file objects to the `Files` structure in the `on_nav` event handler which is not an expected behavior and breaks expectations from other plugins.

Files should be manipulated only on the `on_files` event hander.

The [on_nav](https://github.com/squidfunk/mkdocs-material/blob/master/material/plugins/blog/plugin.py#L139) event handler of the blog plugins calls methods such as `_generate_archive`, `_generate_categories` and `_generate_pages` which are adding file-like objects to the `files` one.

Other plugins such as the `mkdocs-static-i18n` rely on the fact that the `files` structure does not change after the `on_files` event handler is done.

### Related links

- https://github.com/squidfunk/mkdocs-material/blob/master/material/plugins/blog/plugin.py#L139

### Use Cases

Bring i18n support to blogs generated thanks to mkdocs-material

### Before submitting

- [X] I have read and followed the [change request guidelines](https://squidfunk.github.io/mkdocs-material/contributing/requesting-a-change/).
- [X] I have verified that [my idea is a change request and not a bug report](https://squidfunk.github.io/mkdocs-material/contributing/requesting-a-change/#its-not-a-bug-its-a-feature).
- [X] I have ensured that, to the best of my knowledge, [my idea will benefit the entire community](https://squidfunk.github.io/mkdocs-material/contributing/requesting-a-change/#benefit-for-the-community).
- [X] I have included relevant links to [the documentation](https://squidfunk.github.io/mkdocs-material), related [issues](https://github.com/squidfunk/mkdocs-material/issues), and [discussions](https://github.com/squidfunk/mkdocs-material/discussions) to underline the need for my idea.

**Comment by ultrabug** [comment]
In order to unblock the `mkdocs-static-i18n` v1.0.0 release, I intend to move on with https://github.com/ultrabug/mkdocs-static-i18n/pull/241

This makes both plugin compatible as best as currently possible on my part <3

**Comment by squidfunk** [comment]
Thanks for reporting and for making the plugin work with your plugin!

> Other plugins such as the mkdocs-static-i18n rely on the fact that the files structure does not change after the on_files event handler is done.

I agree that it would be favorable to generate all files in `on_files`, but the problem with this is that MkDocs doesn't make this particularly easy because of how navigation is constructed, because MkDocs will always generate either a [`Link` or a `Page` for a file that is not excluded](https://github.com/mkdocs/mkdocs/blob/94e9f17cd7a69e70c18ae282d55ba2f34f93c542/mkdocs/structure/nav.py#L203-L204). As you've probably noticed, the blog plugin makes use of subclasses for pages of specific types, e.g. [`Post`](https://github.com/squidfunk/mkdocs-material/blob/de0c5b7719deff6d092dc673e82af701c6f36001/src/plugins/blog/structure/__init__.py#L48), [`View`](https://github.com/squidfunk/mkdocs-material/blob/de0c5b7719deff6d092dc673e82af701c6f36001/src/plugins/blog/structure/__init__.py#L204), etc., and relations among different concepts in the plugin are managed inside the instances of those subclasses. The problem is, that __MkDocs will always generate a plain old `Page` instance, overwriting the more specific instance that has been generated by the plugin before__. We could still solve it by splitting file generation and page instantiation, but this would make things much more complicated and non-linear, because during generation, [we already assign categories to pages and vice versa](https://github.com/squidfunk/mkdocs-material/blob/de0c5b7719deff6d092dc673e82af701c6f36001/src/plugins/blog/plugin.py#L598-L599), etc. We would need to split this into two stages, blowing up the code base of the plugin and also the runtime, as we're doing two passes. This is clearly not favorable from our perspective.

__Moving forward__: I believe that this problem can be mitigated with a very small but important change in MkDocs: if a `File` already defines a dedicated `Page` in `file.page`, MkDocs should just use that instead of creating a new one:

``` py
return file.page if file.page else Page(title, file, config)
```

It's essentially a single line change in MkDocs that would make things much, much simpler for plugin developers. I've written several plugins now, and I repeatedly had to work around the oddities of this behavior. If you can help facilitate such a change in MkDocs, we should be able to move the entirety of the generation to `on_files`. However, we must ensure that the navigation generation and manipulation in the blog plugin remains as flexible as it currently is.

That being said, we're closing this issue as an upstream issue. If we manage to sort things out over at MkDocs, I encourage you to create a PR that moves generation entirely to `on_files`.

**Comment by squidfunk** [comment]
Good news: I've managed to find a way to move all generated files into `on_files`, leaving `on_nav` only with patching up the navigation and making pagination work. The current design had some further shortcomings which I think I all addressed in [`a0280cb`](https://github.com/squidfunk/mkdocs-material-insiders/commit/a0280cb83dc0bf11af3aa6f6df17d0615eb1ab1f) (Insiders). I will now backport those changes to the community edition, so we have a unified code base. This issue can be considered resolved.

**Comment by ultrabug** [comment]
Congratulations @squidfunk and thanks for getting back to this problem. I have the feeling that you could not live with a `not feasible` tag ;)

Please ping me once it's backported and soon to be released so I can myself adjust to the new behavior.

**Comment by squidfunk** [comment]
It's done already.

**Comment by ultrabug** [comment]
Ok @squidfunk from my first analysis now you generate Page-like objects within on_files :thinking: 

You're going to hate me but so far the resulting structure is worse to work with than before :sob: since there are yielded Page structures with their own files on your `blog.views` mixed with the regular `Files` structure :exploding_head: 

I'll try my luck again later. Meanwhile we can consider our plugins incompatible when you release 9.2.6

**Comment by squidfunk** [comment]
Yes, we're now generating all files and their pages in `on_files`. Your original issue was only concerned regarding files, not pages, which is why I deemed this to be no problem. In my view, it is the absolute correct location to generate files and pages, and as I've written in my last comment, MkDocs should only create `Page` objects for files that don't already have one associated. This would make writing plugins much simpler (cc @oprypin).

The thing is, MkDocs will generate all pages for you between `on_files` and `on_nav` – where, if you use subclassed Page objects like we did (and like others do as well, e.g. the [section-index](https://github.com/oprypin/mkdocs-section-index/blob/f5414c0775848cec96f1181f74c622fe3e14985e/mkdocs_section_index/__init__.py#L12) plugin. Yes, the plugin [monkey-patches the class in `on_nav`](https://github.com/oprypin/mkdocs-section-index/blob/f5414c0775848cec96f1181f74c622fe3e14985e/mkdocs_section_index/plugin.py#L33-L34), same as we did before, but that feels more like a hack than a solution.

> You're going to hate me but so far the resulting structure is worse to work with than before 😭 since there are yielded Page structures with their own files on your blog.views mixed with the regular Files structure 🤯

All generated files are added to the files collection governed by MkDocs, e.g.:

https://github.com/squidfunk/mkdocs-material/blob/df4515c7dbef6bf3790fa16d4435d2b89c85a4a9/src/plugins/blog/plugin.py#L534-L538

The plugin then yields the pages for the caller to put into the correct locations, i.e., interlink views and associate pages to views. I'm not sure what the problem is here – all file are visible to third party plugins after `on_files` and their associated pages can be checked against with `isinstance` for the respective type, as I mentioned in my other comment over at static-i18n. That being said, I'd be interested to hear why we should move page creation out of `on_files` – it's the last chance for us before MkDocs will do it for us prior to `on_nav`. I find it to be the most elegant solution, and if MkDocs would change a single line in the code base, and only create a `Page` if none is present (as mentioned in https://github.com/squidfunk/mkdocs-material/issues/5909#issuecomment-1696958973)), it would make things even simpler.

> I'll try my luck again later. Meanwhile we can consider our plugins incompatible when you release 9.2.6

Feel free to share your progress. I have to release 9.2.6 todays because there are several bugs that need to be fixed, and that are finally addressed by the refactoring I did in the process of moving file generation to `on_files`.
